File: src/main/java/com/smartsoft/csp/varSet/EmptyVarSet.java
package com.smartsoft.csp.varSet;

import com.smartsoft.csp.ast.Ser;
import com.smartsoft.csp.ast.Space;
import com.smartsoft.csp.ast.Var;
import com.smartsoft.csp.parse.VarSpace;
import com.smartsoft.csp.util.ints.IntIterator;

import java.util.NoSuchElementException;

public class EmptyVarSet extends VarSet {

    private static VarSet _INSTANCE;

    public static VarSet getInstance() {
        if (_INSTANCE == null) _INSTANCE = new EmptyVarSet();
        return _INSTANCE;
    }

    private EmptyVarSet() {
    }

    @Override
    public int indexOf(int varId) {
        return -1;
    }

    public VarSpace getVarSpace() {
        throw new UnsupportedOperationException();
    }

    public Space getSpace() {
        throw new UnsupportedOperationException();
    }

    @Override
    public void serialize(Ser a) {

    }

    @Override
    public int getVarSetId() {
        return 0;
    }



    @Override
    public int minVrId() throws NoSuchElementException {
        throw new NoSuchElementException();
    }

    @Override
    public int maxVrId() throws NoSuchElementException {
        throw new NoSuchElementException();
    }

    @Override
    public IntIterator intIterator() {
        return IntIterator.EMPTY;
    }

    @Override
    public boolean containsPrefix(String prefix) {
        return false;
    }

    @Override
    public int size() {
        return 0;
    }

    @Override
    public boolean isEmpty() {
        return true;
    }



//    @Override
//    public boolean anyIntersection(VarSet that) {
//        return false;
//    }

    @Override
    public VarSet immutable() {
        return this;
    }

    @Override
    public int computeContentHash() {
        return 0;
    }

    @Override
    public int getVarId(int index) throws IndexOutOfBoundsException {
        throw new IndexOutOfBoundsException();
    }

    @Override
    public boolean recomputeSize() {
        return false;
    }



}

File: src/main/java/com/smartsoft/csp/varSet/EmptyVarSet.kt
package com.smartsoft.csp.varSet

import com.smartsoft.csp.ast.Ser
import com.smartsoft.csp.ast.Space
import com.smartsoft.csp.parse.VarSpace
import com.smartsoft.csp.util.ints.IntIterator
import java.util.*

object EmptyVarSet : VarSet() {

    override fun getVarSpace(): VarSpace = throw UnsupportedOperationException()

    override fun minVrId(): Int = throw NoSuchElementException()

    override fun maxVrId(): Int = throw NoSuchElementException()

    override fun intIterator(): IntIterator = IntIterator.EMPTY;

    override fun containsPrefix(prefix: String?): Boolean = false

    override fun isEmpty(): Boolean = true

    override fun getVarId(index: Int): Int = throw IndexOutOfBoundsException()


    override fun indexOf(varId: Int): Int = -1

    override fun getSpace(): Space = throw UnsupportedOperationException()

    override fun serialize(a: Ser) {}

    override val size: Int get() = 0

    override fun immutable(): VarSet = this

    override fun computeContentHash(): Int = 0

    @JvmStatic
    fun getInstance(): EmptyVarSet {
        return this
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null) return false
        return other === EmptyVarSet || size == 0
    }
}